_posts/2021-05-18-managing-react-application-state-with-context.md

```diff
@@ -0,0 +1,177 @@
+
+By using Context, we can resolve some of these issues since we are not limited by the opinions enforced by Redux however it does still require some boilerplate code. The benefit of using something built into React is that we can choose to implement it at any time with little to no increase in our overall package size and it is something developers may be familiar with already.
+
+# Using Context for Application State
+
+Using the same principles that Redux was built on, we can manage our application state using something that is built into the React library, namely Context. In terms of developer experience, it is also beneficial for us to use TypeScript which allows us to set up our `State` and `Action` types and benefit from the type checking prior to runtime. In our example, we have a simple application that displays a *count* and allows us to *increment*, *decrement* and *set the value* manually. We begin by setting up each of these actions and a state variable to track the count.
+
+```javascript
+// countProvider.tsx
+type Action =
+  | { type: "increment" }
+  | { type: "decrement" }
+  | { type: "setValue", payload: { value: number } };
+type State = {
+    count: number,
+};
+type Dispatch = (action: Action) => void;
+```
+
+We follow this by setting up our *reducer* function which is exactly the same as it would be if we were using Redux; it is a pure function that takes the state, applies an action to it and returns the new state. I have set up handlers for each of the `Action`s I defined above which will allow us to `increment` and `decrement` the count as well as `setValue` of the count. Notice how I am careful not to directly change the existing state but instead copy the existing state with our changes applied on top of it.
+
+```javascript
+// countProvider.tsx
+function CountReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case "increment":
+      return {
+        ...state,
+        count: state.count + 1,
+      };
+    case "decrement":
+      return {
+        ...state,
+        count: state.count - 1,
+      };
+    case "setValue": {
+      return {
+        ...state,
+        count: action.payload.value,
+      };
+    }
+    default:
+      throw Error(`Unhandled action type: ${action!.type}`);
+  }
+}
+```
+
+Next, I set up Context for both the `State` and `Dispatch` and provide a function component that determines how the `Reducer` we created above will be used. I also provide a way of setting the initial state which aids in testability since you can establish a given state in your tests before performing actions to see how the system behaves.
+
+```javascript
+// countProvider.tsx
+const CountStateContext = createContext<State | undefined>(undefined);
+const CountDispatchContext = createContext<Dispatch | undefined>(undefined);
+
+const CountProvider: FC<{ init?: Partial<State> }> = ({ children, init }) => {
```

Comment:
I didn't even think about it for this one, but also if you pass in `props` to set the initial state and then those `props` change, it won't actually update the state so I feel like that is an indication that you aren't doing the right thing. It will still trigger a re-render though.

Reply:
Also I feel like the name and usage of the `init` might be a good indicator that it shouldn't change once set.